### PR TITLE
Add caching to speed up `bin/packwerk check` and `bin/packwerk update-deprecations` by 85-90%

### PR DESCRIPTION
**Pull request by alexevanczuk**
# Summary
In this PR, I figured out a way that we could add caching to improve runtime performance by 90% when run on all files, getting total runtime down from 3m30 to <20s in our codebase.

I believe this is a better tradeoff than what I tried to do in this PR https://github.com/Shopify/packwerk/pull/167 with a much more complex cache implementation that was hard to maintain and locked us into an implementation.

- The caching is 100% based off of file contents. That is -- we store data that is only a function of the file contents itself. This is because we're caching the result of parsing the file and getting non-local, partially qualified constant definitions. Note that to do this, I pulled some work up higher in the stack -- namely, converting a partially qualified constant to a fully qualified constant. This makes the implementation of the cache *way* simpler mechanically and conceptually.
- The other PR caches a bit more, but a bit of flamegraph analysis was able to capture some more low hanging fruit and get us from 3m30s to 20-30s.
- Note the cache is`200M` (`du -h tmp/cache`). For comparison, the `bootsnap` cache is `99M` and the `sorbet` cache is `148M`. This seems fine to me but open to other perspectives.

# Performance Analysis
I ran `bin/packwerk check` 5 times in our biggest monolith under three scenarios (each with `parallel: true`).
- On `development`
- On this branch, with a cold cache
- On this branch, with a hot cache

Here are the results:
Screen Shot 2021-12-27 at 8 47 04 AM

I did the same with `bin/packwerk check path/to/file.rb`. I picked 100 random ruby files with `Dir['packs/**/*.rb'].sample(100)` and ran a script to run `bin/packwerk check` on those 100 files. For this one I made a box and whiskers diagram to show the distribution more clearly (https://spreadsheetpoint.com/how-to-make-a-box-and-whisker-plot-in-google-sheets/).
Screen Shot 2021-12-27 at 8 38 01 AM
Screen Shot 2021-12-27 at 8 52 20 AM
As you can see this doesn't have a meaningful impact on performance against a single file. There may be some improvements to be made here still, but probably we need step level improvements here (such as switching to a sorbet parser, minimizing the required bundle size, etc.)

# Memory Analysis
Since this cache is used across many processes, the effects of inefficient memory use are multiplied, as seen in https://github.com/Shopify/packwerk/issues/164, so I don't want to exacerbate the issue.

I did an analysis of the memory usage of `bin/packwerk check` before and after this change.
I did this using this simple gem: https://github.com/schneems/get_process_mem
I added call with `process_file`:
```
mem = GetProcessMem.new
puts "Mem: #{mem.mb}"
```

Every time we processed a file, I printed out the memory. I then graphed this over time for before and after to compare. Note I graphed the memory *not* running in parallel. I believe this will be a worst case analysis, because each process would generate a subset of any intermediate cached objects, where a single process would have a larger single cache containing all files. I took three measurements. One is packwerk today with no changes. The other is this PR when the cache is cold, the other is this PR when the cache is warm:
Screen Shot 2021-12-22 at 10 18 16 AM

You can see here that there is no discernable difference in memory usage when the cache is cold, and the memory usage has actually decreased when the cache is hot.

# Flamegraph
I used a flamegraph based analysis to identify performance improvement opportunities (running in serial and without spring so that flamegraph is able to actually correctly sample the process - see the comments in the other PR linked above for how)

The results are that after this, 89% of the time is in finding offenses, 5% in getting load paths (negligible with spring), 1.4% in getting files to process (a second or two, which probably could be improved for the single file check), and 3.3% in bundler related stuff (negligible with spring).

Screen Shot 2021-12-21 at 12 22 09 PM

Of that 89% time, about 60% was spent in `package_from_path`, and 40% was spent in doing YAML loading (the initial approach):
Screen Shot 2021-12-21 at 12 25 28 PM

I was able to reduce the `package_from_path` time considerably using a Trie (private to that class), which shaved off another 20 or so seconds:
Screen Shot 2021-12-21 at 7 19 17 PM
**Edit:** I moved the PackageTrie into a separate PR

Lastly, I changed to use JSON based serialization to get to 13s for our whole monolith. JSON serialization makes sense because the object we are serializing and unserializing is simple and doesn't need all the complexities of YAML.

## What are you trying to accomplish?
Speed up `bin/packwerk check` on all files.

## What approach did you choose and why?
I chose to cache the results of parsing a single file, independent of that file's relation to other files. This I felt like was the best tradeoff of cost/benefit. I tried other approaches in https://github.com/Shopify/packwerk/pull/167

Implementation note:
A previous implementation used YAML.dump and YAML.load.
Although these are much cleaner, the YAML loading code can handle arbitrary object shapes and therefore is a lot more complex (thus, slower). By having a simple serialization and deserialization, we save some time here.

## What should reviewers focus on?
Let me know if there are costs to having this cache (either maintenance costs, performance costs, usability costs, or more) that I'm not seeing.

## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Non-breaking change (a change that doesn't alter functionality - i.e., code refactor, configs, etc.)

### Additional Release Notes

Enable the packwerk check with `EXPERIMENTAL_PACKWERK_CACHE=1 bin/packwerk check` or  `EXPERIMENTAL_PACKWERK_CACHE=1 bin/packwerk update-deprecations`. You can also turn it on globally for all users with `experimental_cache: true` in `packwerk.yml`.

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] It is safe to rollback this change.

### CODE REVIEW

**Comment by technicalpickles** [comment]
This is pretty huge 🔥  Bringing the time to run `packwerk check` down that much makes it feasible to run during our pre-commit hooks.

cc @mclark as someone who is a fan of both packwerk and flamegraphs 👀

**Comment by exterm** [comment]
Super cool investigation!

What's the y axis on your "performance over 100 iterations on a single file" graph?

File: lib/packwerk/cache.rb
Comment (exterm) [inline lib/packwerk/cache.rb:7]: Can we let `packwerk.yml` be part of the cache key, or have a separate cache key for `packwerk.yml` that triggers us to re-parse all source files?

File: lib/packwerk/cache.rb
Comment (exterm) [inline lib/packwerk/cache.rb:8]: This sadly is a little difficult - we could assume all inflections to be defined in a certain file, but we would never cover all cases. For example, we found that some gems add inflections to the main app.

Comment (exterm) [reply lib/packwerk/cache.rb:8]: On the other hand, now that we boot up the application for `packwerk check` anyway, instead of comparing file contents, could we save `application_inflections.hash` and save that, use it as a trigger to clear the cache?

File: lib/packwerk/cache.rb
Comment (exterm) [reply lib/packwerk/cache.rb:13]: However I think we should first evaluate whether we can make it virtually impossible for the misleading behavior to occur.

File: lib/packwerk/package_set.rb
Comment (exterm) [inline lib/packwerk/package_set.rb:191]:
```suggestion
        packages[@package_trie.longest_package_name_that_is_superset_of(filepath_string)]
```

**Comment by exterm** [comment]
As of right now, this works just fine with our monolith.

Cache size is ~300M compared to ~390M bootsnap cache. That seems fine.

Let's figure out the details, specifically around when to invalidate the cache.

File: lib/packwerk/cache.rb
Comment (AlexEvanczuk) [reply lib/packwerk/cache.rb:7]: Sure, I can add `packwerk.yml` as a cache key. Note that only really covers `custom_associations` (exclude/include paths should not affect the output), and changing of inflections would also invalidate the cache.

If we wanted, I suppose, we could keep the JSONified inflections as a cache input too, and invalidate the cache if that jsonified inflections changes 🤔 

That would make this more "complete," although another part of me is thinking of the tradeoff between additional cache complexity vs. practically how it affects user experience (I imagine inflections won't change too often). I'll defer to you on which approach to take! Happy to do either.

File: lib/packwerk/cache.rb
Comment (AlexEvanczuk) [reply lib/packwerk/cache.rb:8]: I wrote my above comment before reading this one. I think conceptually that could work.

That particular implementation I believe would not work:
https://stackoverflow.com/questions/57611448/why-is-ruby-hash-producing-different-output-every-time

Unless `Rails.application_inflections.hash` is implemented differently, `hash` is unstable across processes.

File: lib/packwerk/cache.rb
Comment (AlexEvanczuk) [reply lib/packwerk/cache.rb:13]: I hear ya -- I can see someone getting stuck on this and being annoyed. The cache properly busting would be really nice.

Zooming out -- it sounds like we're saying two things:
1) Let's treat the cache not being busted when it *should* be busted as a bug
2) Let's say these bugs are not good and we should prioritize fixing them (and before they land 😄 )

If you agree with those, I'm game!

File: lib/packwerk/package_set.rb
Comment (AlexEvanczuk) [reply lib/packwerk/package_set.rb:191]: Will change

File: lib/packwerk/cache.rb
Comment (exterm) [reply lib/packwerk/cache.rb:13]: 👍🏼

File: lib/packwerk/cache.rb
Comment (AlexEvanczuk) [reply lib/packwerk/cache.rb]: Here it is! https://github.com/Shopify/packwerk/pull/169/commits/2335703e7ae80a2e79ebe3e851a877261ae8e513

**Review by exterm — APPROVED** [review]

File: USAGE.md
Comment (AlexEvanczuk) [reply USAGE.md:80]: Sounds good to me!

File: USAGE.md
Comment (AlexEvanczuk) [reply USAGE.md:107]: Good point. We originally had it listed as not recommended and experimental because there were some known cases where the cache would become invalid but not properly busted. Now that we've addressed that and are considering issues with the cache as unexpected bugs, we can remove this disclaimer.

File: USAGE.md
Comment (AlexEvanczuk) [reply USAGE.md:104]: Pushing up a commit to remove mentions of "experimental"

File: lib/packwerk/cache.rb
Comment (AlexEvanczuk) [reply lib/packwerk/cache.rb:19]: moved this into the PR description!

File: lib/packwerk/cache.rb
Comment (AlexEvanczuk) [reply lib/packwerk/cache.rb:6]: I didn't realize people would be `cd`ing into a specific package... I wouldn't expect `bundle exec` to succeed unless that package has a `Gemfile` (e.g. it's a gem that also has packwerk installed/maintained). I figured this would only be run through the root. Either way though -- I think having a configurable cache directory makes a ton of sense. Different people may set up their machines differently, and I could see different repos wanting a different cache location, so this is changed!

File: lib/packwerk/file_processor.rb
Comment (AlexEvanczuk) [reply lib/packwerk/file_processor.rb:18]: Yeah -- bad auto format from `Layout/LineLength`. I'll just break over multiple lines.

File: lib/packwerk/package_set.rb
Comment (AlexEvanczuk) [reply lib/packwerk/package_set.rb:10]: I'm reverting this commit so will address this in a follow-up PR!

File: lib/packwerk/package_set.rb
Comment (AlexEvanczuk) [reply lib/packwerk/package_set.rb:11]: Yep, that makes sense!

File: lib/packwerk/reference_extractor.rb
Comment (AlexEvanczuk) [reply lib/packwerk/reference_extractor.rb:28]: fixing

File: test/unit/reference_extractor_test.rb
Comment (AlexEvanczuk) [reply test/unit/reference_extractor_test.rb:251]: fixing

**Comment by AlexEvanczuk** [comment]
@rafaelfranca 
Thank you for all of the great suggestions!

I've responded to each of your suggestions and pushed up new commits to address your concerns.
I pushed up the changes as new commits so that you'd have a pleasant review experience (i.e. you can see what's changed since you last reviewed), but if you're ready for this to land, I can merge those commits into the previous ones so that what lands on `main` is clean.

**Comment by rafaelfranca** [comment]
Great work on this. I'll try this branch in our application just to make sure it is working.

**Pull request merged**
